app/models/proposal.rb

```diff
@@ -92,43 +92,31 @@ def users
     results.compact
   end
 
-  # returns the Approval
-  def add_approver(email)
-    user = User.for_email(email)
-    approval = self.approvals.create!(user_id: user.id)
-    approval
-  end
-
-  def remove_approver(email)
-    user = User.for_email(email)
-    approval = self.existing_approval_for(user)
-    approval.destroy
-  end
-
-  # Set the approver list, from any start state
-  # This overrides the `through` relation but provides parity to the accessor
-  def approvers=(approver_list)
-    approvals = approver_list.each_with_index.map do |approver, idx|
-      approval = self.existing_approval_for(approver)
-      approval ||= Approval.new(user: approver, proposal: self)
-      approval.position = idx + 1   # start with 1
-      approval
+  # Sets the approval list from any start state, reusing any user approvals
+  def create_or_update_approvals(new_approvals)
+    new_approvals = new_approvals.each_with_index.map do |new_approval, idx|
+      user = new_approval.user
+      if user && existing = self.existing_approval_for(user)
+        existing.parent = new_approval.parent   # this assumes the parent hasn't been replaced
+        existing
+      else
+        new_approval
+      end
+    end
+    self.approvals = new_approvals
+    # position might be out of whack, so we reset it
+    new_approvals.each_with_index do |approval, idx|
+      approval.set_list_position(idx + 1)   # start with 1
```

Comment:
As the approvals are potentially being modified, we'll want to add an explicit field for approval approved time.